lib/puppet/configurer.rb

```diff
@@ -397,22 +397,32 @@ def run_internal(options)
       if !cached_catalog && options[:catalog]
         ral_catalog = options[:catalog]
       else
-        # If not noop, commit the cached resource catalog (not ral catalog). Ideally
+        # Ordering here matters. We have to resolve deferred resources in the
+        # resource catalog, convert the resource catalog to a RAL catalog (which
+        # triggers type/provider validation), and only if that is successful,
+        # should we cache the *original* resource catalog. However, deferred
+        # evaluation mutates the resource catalog, so we need to make a copy of
+        # it here. If PUP-9323 is ever implemented so that we resolve deferred
+        # resources in the RAL catalog as they are needed, then we could eliminate
+        # this step.
+        catalog_to_cache = Puppet.override(:rich_data => Puppet[:rich_data]) do
+          Puppet::Resource::Catalog.from_data_hash(catalog.to_data_hash)
+        end
+
+        # REMIND @duration is the time spent loading the last catalog, and doesn't
+        # account for things like we failed to download and fell back to the cache
+        ral_catalog = convert_catalog(catalog, @duration, facts, options)
+
+        # Validation succeeded, so commit the `catalog_to_cache` for non-noop runs. Don't
+        # commit `catalog` since it contains the result of deferred evaluation. Ideally
         # we'd just copy the downloaded response body, instead of serializing the
         # in-memory catalog, but that's hard due to the indirector.
-        #
-        # Important to save the catalog before evaluating deferred functions, as we
-        # want to cache the deferred function, not its result.
         indirection = Puppet::Resource::Catalog.indirection
         if !Puppet[:noop] && indirection.cache?
-          request = indirection.request(:save, nil, catalog, environment: Puppet::Node::Environment.remote(catalog.environment))
+          request = indirection.request(:save, nil, catalog_to_cache, environment: Puppet::Node::Environment.remote(catalog_to_cache.environment))
           Puppet.info("Caching catalog for #{request.key}")
           indirection.cache.save(request)
         end
```

Comment:
It also seems like we should only cache the catalog if it was successfully applied (after https://github.com/puppetlabs/puppet/pull/8479/files#diff-51e52bffe1de00c153bde14dea475f15abdcb6a1fdaf2691633dad8238e654cfR433). But that's a bigger change than this PR is fixing (resolving the regression introduced in the initial fix for PUP-10818).